utils/src/main/java/software/amazon/awssdk/utils/internal/MappingSubscriber.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.internal;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Function;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * Maps a subscriber of one type to another type. If an exception is thrown by the mapping function itself, the error
+ * will be propagated to the downstream subscriber as if it had come from the publisher and then the subscription will
+ * be implicitly cancelled and no further events from the publisher will be passed along.
+ */
+@SdkInternalApi
+public class MappingSubscriber<T, U> implements Subscriber<T> {
+
+    private final Subscriber<? super U> delegateSubscriber;
+    private final Function<T, U> mapFunction;
+    private boolean isCancelled = false;
+    private final Set<Subscription> subscriptions = new HashSet<>();
+
+    private MappingSubscriber(Subscriber<? super U> delegateSubscriber,
+                              Function<T, U> mapFunction) {
+        this.delegateSubscriber = delegateSubscriber;
+        this.mapFunction = mapFunction;
+    }
+
+    public static <T, U> MappingSubscriber<T, U> create(Subscriber<? super U> subscriber,
+                                                        Function<T, U> mapFunction) {
+        return new MappingSubscriber<>(subscriber, mapFunction);
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        this.subscriptions.add(subscription);
+        delegateSubscriber.onSubscribe(subscription);
+    }
```

Comment:
I'm not sure about this; any reason we're supporting multiple subscriptions? Per my reading of the spec, a `Subscriber` can only have one `Subscription`. Per [2.12](https://github.com/reactive-streams/reactive-streams-jvm/tree/v1.0.3#2.12):

> Subscriber.onSubscribe MUST be called at most once for a given Subscriber (based on object equality).